### PR TITLE
docs: add documents for the data structure of Cell/Script/Transaction/Block

### PR DESCRIPTION
**Pull request by Mine77**
#308 
This document explains "What does each field mean?"
As for "Who is responsible to fill them?" and "How to generate them?" , I think it's more appropriate to document them it in another place, which I'm also working on in progress.

[preview](https://github.com/Mine77/ckb/blob/docs/add-data-structure/docs/data-structures.md)

### CODE REVIEW

File: docs/data-structures.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    "version": 0
+}
+```
+
+### Description
+
+| Name                     | Type         | Description                                                  |
+| ------------------------ | ------------ | ------------------------------------------------------------ |
+| `version`                | uint32       | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `deps`                   | [`cell`]     | **An array of cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |
```
Comment (Keith-CY) [inline docs/data-structures.md]: Type is `[out_point]`

File: docs/data-structures.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+| `inputs.previous_output` | [`outpoint`] | **An array of cell outpoints that point to the cells used as inputs.** Input cells are in fact the output of previous transactions, hence they are noted as `previous_output` here. These cells are referred through  `outpoint`, which contains the transaction `hash` of the previous transaction, as well as this cell's `index` in its transaction's output list. |
```
Comment (Keith-CY) [inline docs/data-structures.md]: `inputs` is an array of  `{previous_output, unlock}`

File: docs/data-structures.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    "lock": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
+    "type": null
+}
+```
+
+## Description
+
+| Name       | Type       | Description                                                  |
+| :--------- | :--------- | :----------------------------------------------------------- |
+| `capacity` | uint64     | **The size of the cell.** When a new cell is generated (via transaction), one of the verification rule is `capacity = len(capacity)+len(data)+len(type)+len(lock)`. This value also represents the balance of CKB coin, just like the `balance` field in the Bitcoin's UTXO. (E.g. Alice owns 100 CKB coins means she can unlock a group of cells that has 100 amount of `capacity` in total.) |
```
Comment (doitian) [inline docs/data-structures.md]: The rule is `capacity ≥ len(capacity)+len(data)+len(type)+len(lock)`

File: docs/data-structures.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+| `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logics and rules defined by `type` or `lock`. |
+| `type`     | `Script`   | **A Script that defines the type of the cell.** In a transaction, if an input cell and an output cell has the same `type` field, then the `data` part of these two cells is limited by the `type` script upon the transaction verification. (I.e. `type` is a script that limits how the `data` field of the new cells can be changed from the old cells.) `type` is required to has a data structure of `script`. This field can be empty in a Cell. |
```
Comment (doitian) [inline docs/data-structures.md]: The description is not accurate.

@xxuejie Could you help to give a description of `type`?

File: docs/data-structures.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+
+
+### Description
+
+| Name          | Type    | Description                                                  |
+| :------------ | :------ | :----------------------------------------------------------- |
+| `version`     | uint8   | **The version of the script.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `binary`      | Bytes   | **ELF formatted binary that contains an RISC-V based script.** This part of data is loaded into an CKB-VM instance when they are specified upon the transaction verification. |
+| `reference`   | Bytes   | **The hash of the script that is referred by this script.** It is possible to refer the script in another cell on-chain as the binary code in this script, instead of entering the binary directly into the script. **Notice:** This is part only works when the `binary` field is empty. |
```
Comment (doitian) [inline docs/data-structures.md]: How the cell is referred to here?

@xxuejie can provide help here.

File: docs/data-structures.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+| Name              | Type                             | Description                                                  |
+| ----------------- | -------------------------------- | ------------------------------------------------------------ |
+| `version`         | uint32                           | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `deps`            | [`outpoint`]                     | **An array of `outpoint` that point to the cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |
+| `inputs`          | [{`previsou_output` , `unlock`}] | **An array of {`previsou_output` , `unlock`}.**              |
+| `previous_output` | [`outpoint`]                     | **An array of cell outpoints that point to the cells used as inputs.** Input cells are in fact the output of previous transactions, hence they are noted as `previous_output` here. These cells are referred through  `outpoint`, which contains the transaction `hash` of the previous transaction, as well as this cell's `index` in its transaction's output list. |
```
Comment (doitian) [inline docs/data-structures.md]: `previous_output` is a `outpoint`, not an array of outpoints

File: docs/data-structures.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+| `unlock`          | [`script`]                       | **An array of scripts for unlocking their related input cells** (i.e. `previous_output`). See [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md) for how to program this part. |
```
Comment (doitian) [inline docs/data-structures.md]: `unlock` is a `script`, not an array.

File: docs/data-structures.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+| `header`                | `Header`        | **The block header of the block.** This part contains some metadata of the block. |
+| `commit_trasactions`    | [`Transaction`] | **An array of transactions contained in the block.** This is where the miner put the received transactions. |
+| `proposal_transactions` | [string]        | **An array of hex-encoded short transaction ID.**            |
+| `uncles`                | [`UncleBlock`]  | **An array of uncle blocks of the block.**                   |
+
+#### Header
+
+| Name           | Type                | Description                                                  |
+| -------------- | ------------------- | ------------------------------------------------------------ |
+| `cellbase_id`  | H256(hash)          | **The hash of the Cellbase transaction.** Cellbase transaction is just like the coinable transaction in Bitcoin. It's the transaction added by the miner who mined this block, by which the miner receives block reward for successfully mined the block. |
```
Comment (doitian) [inline docs/data-structures.md]:
```suggestion
| `cellbase_id`  | H256(hash)          | **The hash of the Cellbase transaction.** Cellbase transaction is just like the coinbase transaction in Bitcoin. It's the transaction added by the miner who mined this block, by which the miner receives block reward for successfully mined the block. |
```

File: docs/data-structures.md [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+| `difficulty`   | Bytes               | **The difficulty of the PoW puzzle.**                        |
+| `hash`         | H256(hash)          | **The block hash.**                                          |
```
Comment (doitian) [inline docs/data-structures.md]: RPC returns this, but indeed it is not a field of the header, it is computed from the header.

File: docs/data-structures.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+| `parent_hash`  | H256(hash)          | **The hash of the parent block.**                            |
+| `seal`         | `nonce` and `proof` | **The seal of a block.** After finished the block assembling, the miner can start to do the calculation for finding the solution of the PoW puzzle. The "solution" here is called `seal`. |
+| `seal.nonce`   | uint64              | **The nonce.** Similar to [the nonce in Bitcoin](https://en.bitcoin.it/wiki/Nonce). |
+| `seal.proof`   | Bytes               | **The solution of the PoW puzzle.**                          |
+| `timestamp`    | uint64              | **A [Unix time](http://en.wikipedia.org/wiki/Unix_time) timestamp.** |
+| `txs_commit`   | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of transactions as leaves.** |
+| `txs_proposal` | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of short transaction IDs as leaves.** |
+| `uncles_count` | uint32              | **The number of uncle blocks.**                              |
+| `uncles_hash`  | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of uncle blocks as leaves.** |
```
Comment (doitian) [inline docs/data-structures.md]: It is not the Merkle Root now.

@zhangsoledad Please provide a description for this field.

File: docs/data-structures.md [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+    "type": null
+}
+```
+
+## Description
+
+| Name       | Type       | Description                                                  |
+| :--------- | :--------- | :----------------------------------------------------------- |
+| `capacity` | uint64     | **The size of the cell.** When a new cell is generated (via transaction), one of the verification rule is `capacity ≥ len(capacity)+len(data)+len(type)+len(lock)`. This value also represents the balance of CKB coin, just like the `balance` field in the Bitcoin's UTXO. (E.g. Alice owns 100 CKB coins means she can unlock a group of cells that has 100 amount of `capacity` in total.) |
+| `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logics and rules defined by `type` or `lock`. |
```
Comment (u2) [inline docs/data-structures.md]:
```suggestion
| `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logic and rules defined by `type`. |
```

The data field is only constrained by the `type`, The `lock` defines the ownership of the cell.

File: docs/data-structures.md [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+| `type`     | `Script`   | **A Script that defines the type of the cell.** In a transaction, if an input cell and an output cell has the same `type` field, then the `data` part of these two cells is limited by the `type` script upon the transaction verification. (I.e. `type` is a script that limits how the `data` field of the new cells can be changed from the old cells.) `type` is required to has a data structure of `script`. This field can be empty in a Cell. |
+| `lock`     | H256(hash) | **The hash of a Script that defines the ownership of the cell**, just like the `lock` field in the Bitcoin's UTXO. Whoever can provide an unlock script that has the same hash of a cell's `lock` hash can use this cell as input in an transaction (i.e. has the ownership of this cell). This is similar to the P2SH scheme in Bitcoin. |
+
+
+
+More information about Cell can be found in the [whitepaper](https://github.com/nervosnetwork/rfcs/blob/afe50463bb620393b179bd8f08c263b78e366ab3/rfcs/0002-ckb/0002-ckb.md#42-cell).
```
Comment (u2) [inline docs/data-structures.md]:
```suggestion
More information about Cell can be found in the [whitepaper](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0002-ckb/0002-ckb.md#42-cell).
```

File: docs/data-structures.md (hunk 4 above)
Comment (u2) [reply docs/data-structures.md]: `type` is an option field, it should be explicit to explain.

File: docs/data-structures.md (hunk 5 above)
Comment (u2) [reply docs/data-structures.md]: https://github.com/nervosnetwork/ckb/blob/develop/core/src/script.rs#L13

File: docs/data-structures.md [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+| `version`     | uint8   | **The version of the script.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
+| `binary`      | Bytes   | **ELF formatted binary that contains an RISC-V based script.** This part of data is loaded into an CKB-VM instance when they are specified upon the transaction verification. |
+| `reference`   | Bytes   | **The hash of the script that is referred by this script.** It is possible to refer the script in another cell on-chain as the binary code in this script, instead of entering the binary directly into the script. **Notice:** This is part only works when the `binary` field is empty. |
+| `args`        | [Bytes] | **An array of arguments for the script.** The arguments here are imported into the CKB-VM instance as input arguments for the scripts. This part is NOT used when calculating the hash of the script. |
+| `signed_args` | [Bytes] | **An array of signed arguments for the script.** The arguments with signatures. The `signed_args` and the `args` will be connected into a single vector, and imported into the CKB-VM instance as input arguments. |
```
Comment (u2) [inline docs/data-structures.md]: Most of the time, we put `pubkey` in the `signed_args` and put the signature in `args`. The description of `The arguments with signatures.` has made me confused before, I'm not sure it's clear enough.

File: docs/data-structures.md [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+
+
+
+More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/27669cd6b4f56f8977b725fc0e6582b288ab2b82/docs/how-to-write-contracts.md#script-model).
```
Comment (u2) [inline docs/data-structures.md]:
```suggestion
More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md#script-model).
```

File: docs/data-structures.md [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+}
+```
+
+### Description
+
+#### Transaction
+
+| Name              | Type                             | Description                                                  |
+| ----------------- | -------------------------------- | ------------------------------------------------------------ |
+| `version`         | uint32                           | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
```
Comment (u2) [inline docs/data-structures.md]:
```suggestion
| `version`         | uint32                           | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain. |
```

File: docs/data-structures.md [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+| `deps`            | [`outpoint`]                     | **An array of `outpoint` that point to the cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |
+| `inputs`          | [{`previsou_output` , `unlock`}] | **An array of {`previsou_output` , `unlock`}.**              |
+| `previous_output` | `outpoint`                       | **A cell outpoint that point to the cells used as inputs.** Input cells are in fact the output of previous transactions, hence they are noted as `previous_output` here. These cells are referred through  `outpoint`, which contains the transaction `hash` of the previous transaction, as well as this cell's `index` in its transaction's output list. |
+| `unlock`          | `script`                         | **A script for unlocking the corresponding input cell** (i.e. `previous_output`). See [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md) for how to program this part. |
```
Comment (u2) [inline docs/data-structures.md]: The `previous_output` and `unlock` should not be included, they are part of `inputs`, not the transaction.

File: docs/data-structures.md [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+| ---------------- | ------------------ | ------------------------------------------------------------ |
+| `outpoint`       | {`hash` , `index`} | **An outpoint is pointer to a specific cell.** This is used in a transaction to refer a cell that is generated in a previous transaction. |
+| `outpoint.hash`  | H256(hash)         | **The hash of the transaction that this cell belongs to.**   |
+| `outpoint.index` | uint32             | **The index of the cell in its transaction's output list.**  |
+
+
+
+
+
+More information about the Transaction of Nervos CKB can be found in [whitepaper](https://github.com/nervosnetwork/rfcs/blob/afe50463bb620393b179bd8f08c263b78e366ab3/rfcs/0002-ckb/0002-ckb.md#44-transaction).
```
Comment (u2) [inline docs/data-structures.md]:
```suggestion
More information about the Transaction of Nervos CKB can be found in [whitepaper](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0002-ckb/0002-ckb.md#44-transaction).
```

File: docs/data-structures.md (hunk 10 above)
Comment (Mine77) [reply docs/data-structures.md]: The hash of serialized uncle blocks data?

File: docs/data-structures.md [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+| `difficulty`   | Bytes               | **The difficulty of the PoW puzzle.**                        |
+| `hash`         | H256(hash)          | **The block hash.** Note that although this value is returned by the JSON-RPC API, it is in fact NOT a part of the header. This value is actually calculated from the `Header` data. |
+| `number`       | uint64              | **The block height.**                                        |
+| `parent_hash`  | H256(hash)          | **The hash of the parent block.**                            |
+| `seal`         | `nonce` and `proof` | **The seal of a block.** After finished the block assembling, the miner can start to do the calculation for finding the solution of the PoW puzzle. The "solution" here is called `seal`. |
+| `seal.nonce`   | uint64              | **The nonce.** Similar to [the nonce in Bitcoin](https://en.bitcoin.it/wiki/Nonce). |
+| `seal.proof`   | Bytes               | **The solution of the PoW puzzle.**                          |
```
Comment (u2) [inline docs/data-structures.md]: It's better to create a new table for `Seal` type.

File: docs/data-structures.md (hunk 13 above)
Comment (doitian) [reply docs/data-structures.md]: Yes, *The arguments with signatures* is confusing, they are indeed *arguments included when computing signatures*.

File: docs/data-structures.md (hunk 10 above)
Comment (doitian) [reply docs/data-structures.md]: This looks good to me.

File: docs/data-structures.md (hunk 11 above)
Comment (Mine77) [reply docs/data-structures.md]: It also could be constrained by `lock`, as `type` and `lock` both has access to all the information of the transaction.

File: docs/data-structures.md (hunk 13 above)
Comment (Mine77) [reply docs/data-structures.md]: I explained this in [another doc](https://github.com/Mine77/ckb-demo-ruby-sdk/blob/docs/update-docs/docs/how-to-write-contracts.md#script-model), maybe I should explain less here and just refer this doc directly.

File: docs/data-structures.md (hunk 16 above)
Comment (Mine77) [reply docs/data-structures.md]: The fact that `inputs` is part of transaction and `previsou_output` and `unlock` is part of `input` makes them a part of `inputs` as well.

A keyboard is a part of a laptop, which does not make any of the keys is not part of it.

So I don't think this part could cause too much confusion.

More importantly, this docs serve as a dictionary for looking up parameters, its more convenient  to put related things near each other.

File: docs/data-structures.md (hunk 18 above)
Comment (Mine77) [reply docs/data-structures.md]: I don think it's necessary, same reason to the `input`.

It only increase the complexity.

File: docs/data-structures.md (hunk 15 above)
Comment (Mine77) [reply docs/data-structures.md]: "blockchain" is just a data structure. 

"blockchain system" is the system that utilize a blockchain data structure.

File: docs/data-structures.md (hunk 4 above)
Comment (xxuejie) [reply docs/data-structures.md]: `A script used to enforce certain cell properties. Typical use cases including user defined token support, cell data format validation, etc.`

Comment (Mine77) [reply docs/data-structures.md]: done

File: docs/data-structures.md (hunk 10 above)
Comment (Mine77) [reply docs/data-structures.md]: done

File: docs/data-structures.md [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+| `binary`      | Bytes   | **ELF formatted binary that contains an RISC-V based script.** This part of data is loaded into an CKB-VM instance when they are specified upon the transaction verification. |
+| `reference`   | Bytes   | **The `type hash` of the script that is referred by this script.** It is possible to refer the script in another cell on-chain as the binary code in this script, instead of entering the binary directly into the script. **Notice:** This is part only works when the `binary` field is empty. |
+| `args`        | [Bytes] | **An array of arguments as the script input.** The arguments here are imported into the CKB-VM instance as input arguments for the scripts. This part is NOT used when calculating the hash of the script. |
+| `signed_args` | [Bytes] | **An array of arguments that belongs to the script for improving code reuse rate**. Please refer [this document](https://github.com/Mine77/ckb-demo-ruby-sdk/blob/docs/update-docs/docs/how-to-write-contracts.md#script-model) for more explanation about this field. |
```
Comment (xxuejie) [inline docs/data-structures.md]: This sounds weird to me(`improving code reuse` part), but I don't have a good suggestion, any thoughts @doitian ?

File: docs/data-structures.md [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+
+
+
+More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/master/docs/how-to-write-contracts.md#script-model). Also you can find how the `Script` structure is implemented by reading [these codes](https://github.com/nervosnetwork/ckb/blob/master/core/src/script.rs#L13).
```
Comment (xxuejie) [inline docs/data-structures.md]: What about using links like this: https://github.com/nervosnetwork/ckb/blob/17b48a0777753a91691146b3817758c02db6f415/core/src/script.rs#L13 to ensure future changes won't put this link to a line that is not relevant?

File: docs/data-structures.md (hunk 19 above)
Comment (doitian) [reply docs/data-structures.md]: If it is weird, I would suggest just delete it until we have a better understanding and a brief description available.

File: docs/data-structures.md (hunk 11 above)
Comment (u2) [reply docs/data-structures.md]: I see.

File: docs/data-structures.md (hunk 19 above)
Comment (Mine77) [reply docs/data-structures.md]: I think it's ok to leave it this way.

At least it's just weird to you, not causing confusions. 

It indeed is increasing the code reuse rate.

File: docs/data-structures.md [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+| `parent_hash`  | H256(hash)          | **The hash of the parent block.**                            |
+| `seal`         | `nonce` and `proof` | **The seal of a block.** After finished the block assembling, the miner can start to do the calculation for finding the solution of the PoW puzzle. The "solution" here is called `seal`. |
+| `seal.nonce`   | uint64              | **The nonce.** Similar to [the nonce in Bitcoin](https://en.bitcoin.it/wiki/Nonce). |
+| `seal.proof`   | Bytes               | **The solution of the PoW puzzle.**                          |
+| `timestamp`    | uint64              | **A [Unix time](http://en.wikipedia.org/wiki/Unix_time) timestamp.** |
+| `txs_commit`   | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of transactions as leaves.** |
+| `txs_proposal` | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of short transaction IDs as leaves.** |
+| `uncles_count` | uint32              | **The number of uncle blocks.**                              |
+| `uncles_hash`  | H256(hash)          | **The hash of the serialized uncle blocks data.** This will later be changed to using Merle Trie as well. |
```
Comment (zhangsoledad) [inline docs/data-structures.md]: The serialization will be changed. Will not use Merle Trie.

**Comment by Mine77** [comment]
Can we push this forward?
@doitian

File: docs/data-structures.md (hunk 21 above)
Comment (Mine77) [reply docs/data-structures.md]: thx!
done

**Pull request merged**